File: src/blind-box-leaderboard.ts
import type { ContributionLedger, ViewerContribution } from './types';

export interface BlindBoxLeaderboardSummary {
  viewerCount: number;
  blindBoxCount: number;
  cost: number;
  value: number;
  profit: number;
  unpricedCount: number;
}

export interface BlindBoxLeaderboard {
  summary: BlindBoxLeaderboardSummary;
  viewers: ViewerContribution[];
}

export interface BlindBoxLeaderboardScope {
  giftId: number;
  giftName: string;
  count: number;
  lastGiftAt: number;
}

/**
 * Builds the canonical blind-box leaderboard from the backend-owned contribution ledger.
 * Positive profit means the viewer opened more gift value than the blind boxes cost.
 */
export function buildBlindBoxLeaderboard(
  ledger: ContributionLedger,
  limit = Number.POSITIVE_INFINITY,
  blindBoxGiftId?: number,
): BlindBoxLeaderboard {
  const eligible = ledger.viewers.flatMap((viewer) => {
    if (!blindBoxGiftId) return finite(viewer.blindBoxCount) > 0 ? [viewer] : [];
    const breakdown = viewer.blindBoxes?.find((candidate) => candidate.giftId === blindBoxGiftId);
    if (!breakdown || finite(breakdown.count) <= 0) return [];
    return [{
      ...viewer,
      blindBoxCount: finite(breakdown.count),
      blindBoxCost: finite(breakdown.cost),
      blindBoxValue: finite(breakdown.value),
      blindBoxProfit: finite(breakdown.value) - finite(breakdown.cost),
      unpricedBlindBoxCount: finite(breakdown.unpricedCount),
      lastGiftAt: finite(breakdown.lastGiftAt),
    }];
  });
  const viewers = [...eligible].sort((left, right) => (
    finite(right.blindBoxProfit) - finite(left.blindBoxProfit)
    || finite(right.blindBoxValue) - finite(left.blindBoxValue)
    || finite(right.blindBoxCount) - finite(left.blindBoxCount)
    || finite(right.lastGiftAt) - finite(left.lastGiftAt)
  ));
  const safeLimit = Number.isFinite(limit) ? Math.max(0, Math.floor(limit)) : viewers.length;
  const summary = eligible.reduce<BlindBoxLeaderboardSummary>((result, viewer) => ({
    viewerCount: result.viewerCount + 1,
    blindBoxCount: result.blindBoxCount + finite(viewer.blindBoxCount),
    cost: result.cost + finite(viewer.blindBoxCost),
    value: result.value + finite(viewer.blindBoxValue),
    profit: result.profit + finite(viewer.blindBoxProfit),
    unpricedCount: result.unpricedCount + finite(viewer.unpricedBlindBoxCount),
  }), {
    viewerCount: 0,
    blindBoxCount: 0,
    cost: 0,
    value: 0,
    profit: 0,
    unpricedCount: 0,
  });
  return { summary, viewers: viewers.slice(0, safeLimit) };
}

export function listBlindBoxLeaderboardScopes(ledger: ContributionLedger): BlindBoxLeaderboardScope[] {
  const scopes = new Map<number, BlindBoxLeaderboardScope>();
  for (const viewer of ledger.viewers) {
    for (const breakdown of viewer.blindBoxes ?? []) {
      const giftId = Math.trunc(finite(breakdown.giftId));
      const count = finite(breakdown.count);
      if (giftId <= 0 || count <= 0) continue;
      const giftName = String(breakdown.giftName ?? '').trim() || `盲盒 ${giftId}`;
      const lastGiftAt = finite(breakdown.lastGiftAt);
      const current = scopes.get(giftId);
      if (current) {
        current.count += count;
        if (lastGiftAt >= current.lastGiftAt) {
          current.giftName = giftName;
          current.lastGiftAt = lastGiftAt;
        }
      } else {
        scopes.set(giftId, { giftId, giftName, count, lastGiftAt });
      }
    }
  }
  return Array.from(scopes.values()).sort((left, right) => (
    right.count - left.count
    || right.lastGiftAt - left.lastGiftAt
    || left.giftName.localeCompare(right.giftName, 'zh-CN')
  ));
}

function finite(value: number | undefined): number {
  const normalized = Number(value ?? 0);
  return Number.isFinite(normalized) ? normalized : 0;
}

File: goserver/blind_box_leaderboard.go
package main

import (
	"math"
	"sort"
	"strconv"
	"strings"

	"golang.org/x/text/collate"
	"golang.org/x/text/language"
)

type blindBoxLeaderboardQuery struct {
	GiftID   int
	Limit    int
	HasLimit bool
}

type blindBoxLeaderboardSummary struct {
	ViewerCount   int     `json:"viewerCount"`
	BlindBoxCount int     `json:"blindBoxCount"`
	Cost          float64 `json:"cost"`
	Value         float64 `json:"value"`
	Profit        float64 `json:"profit"`
	UnpricedCount int     `json:"unpricedCount"`
}

type blindBoxLeaderboardScope struct {
	GiftID     int    `json:"giftId"`
	GiftName   string `json:"giftName"`
	Count      int    `json:"count"`
	LastGiftAt int64  `json:"lastGiftAt"`
}

type blindBoxLeaderboardSnapshot struct {
	UpdatedAt int64                      `json:"updatedAt"`
	Summary   blindBoxLeaderboardSummary `json:"summary"`
	Viewers   []viewerContribution       `json:"viewers"`
	Scopes    []blindBoxLeaderboardScope `json:"scopes"`
}

func buildBlindBoxLeaderboard(ledger contributionLedgerState, query blindBoxLeaderboardQuery) blindBoxLeaderboardSnapshot {
	snapshot := blindBoxLeaderboardSnapshot{
		UpdatedAt: normalizeBlindBoxTimestamp(ledger.UpdatedAt),
		Viewers:   []viewerContribution{},
		Scopes:    buildBlindBoxLeaderboardScopes(ledger),
	}

	for _, viewer := range ledger.Viewers {
		projected, ok := projectBlindBoxViewer(viewer, query.GiftID)
		if !ok {
			continue
		}
		snapshot.Viewers = append(snapshot.Viewers, projected)
		snapshot.Summary.ViewerCount++
		snapshot.Summary.BlindBoxCount += projected.BlindBoxCount
		snapshot.Summary.Cost += projected.BlindBoxCost
		snapshot.Summary.Value += projected.BlindBoxValue
		snapshot.Summary.Profit += projected.BlindBoxProfit
		snapshot.Summary.UnpricedCount += projected.UnpricedBlindBoxCount
	}

	sort.SliceStable(snapshot.Viewers, func(left, right int) bool {
		return compareBlindBoxViewers(snapshot.Viewers[left], snapshot.Viewers[right]) < 0
	})

	if query.HasLimit {
		limit := max(0, query.Limit)
		if limit < len(snapshot.Viewers) {
			snapshot.Viewers = snapshot.Viewers[:limit]
		}
	}
	return snapshot
}

func projectBlindBoxViewer(viewer viewerContribution, giftID int) (viewerContribution, bool) {
	if giftID > 0 {
		for _, breakdown := range viewer.BlindBoxes {
			if breakdown.GiftID != giftID {
				continue
			}
			return projectBlindBoxBreakdown(viewer, breakdown)
		}
		return viewerContribution{}, false
	}

	projected := viewer
	projected.BlindBoxCount = normalizeBlindBoxCount(viewer.BlindBoxCount)
	if projected.BlindBoxCount == 0 {
		return viewerContribution{}, false
	}
	projected.BlindBoxCost = normalizeBlindBoxAmount(viewer.BlindBoxCost)
	projected.BlindBoxValue = normalizeBlindBoxAmount(viewer.BlindBoxValue)
	projected.BlindBoxProfit = projected.BlindBoxValue - projected.BlindBoxCost
	projected.UnpricedBlindBoxCount = normalizeBlindBoxCount(viewer.UnpricedBlindBoxCount)
	projected.LastGiftAt = normalizeBlindBoxTimestamp(viewer.LastGiftAt)
	return projected, true
}

func projectBlindBoxBreakdown(viewer viewerContribution, breakdown blindBoxContribution) (viewerContribution, bool) {
	if breakdown.GiftID <= 0 || breakdown.Count <= 0 {
		return viewerContribution{}, false
	}

	projected := viewer
	projected.BlindBoxCount = normalizeBlindBoxCount(breakdown.Count)
	if projected.BlindBoxCount == 0 {
		return viewerContribution{}, false
	}
	projected.BlindBoxCost = normalizeBlindBoxAmount(breakdown.Cost)
	projected.BlindBoxValue = normalizeBlindBoxAmount(breakdown.Value)
	projected.BlindBoxProfit = projected.BlindBoxValue - projected.BlindBoxCost
	projected.UnpricedBlindBoxCount = normalizeBlindBoxCount(breakdown.UnpricedCount)
	projected.LastGiftAt = normalizeBlindBoxTimestamp(breakdown.LastGiftAt)
	return projected, true
}

func buildBlindBoxLeaderboardScopes(ledger contributionLedgerState) []blindBoxLeaderboardScope {
	byGiftID := make(map[int]blindBoxLeaderboardScope)
	for _, viewer := range ledger.Viewers {
		for _, breakdown := range viewer.BlindBoxes {
			if breakdown.GiftID <= 0 || breakdown.Count <= 0 {
				continue
			}
			count := normalizeBlindBoxCount(breakdown.Count)
			if count == 0 {
				continue
			}
			giftName := strings.TrimSpace(breakdown.GiftName)
			if giftName == "" {
				giftName = "盲盒 " + strconv.Itoa(breakdown.GiftID)
			}
			lastGiftAt := normalizeBlindBoxTimestamp(breakdown.LastGiftAt)
			current, exists := byGiftID[breakdown.GiftID]
			if !exists {
				byGiftID[breakdown.GiftID] = blindBoxLeaderboardScope{
					GiftID: breakdown.GiftID, GiftName: giftName, Count: count, LastGiftAt: lastGiftAt,
				}
				continue
			}
			current.Count += count
			if lastGiftAt >= current.LastGiftAt {
				current.GiftName = giftName
				current.LastGiftAt = lastGiftAt
			}
			byGiftID[breakdown.GiftID] = current
		}
	}

	scopes := make([]blindBoxLeaderboardScope, 0, len(byGiftID))
	for _, scope := range byGiftID {
		scopes = append(scopes, scope)
	}
	collator := collate.New(language.Chinese)
	sort.Slice(scopes, func(left, right int) bool {
		if scopes[left].Count != scopes[right].Count {
			return scopes[left].Count > scopes[right].Count
		}
		if scopes[left].LastGiftAt != scopes[right].LastGiftAt {
			return scopes[left].LastGiftAt > scopes[right].LastGiftAt
		}
		if nameOrder := collator.CompareString(scopes[left].GiftName, scopes[right].GiftName); nameOrder != 0 {
			return nameOrder < 0
		}
		return scopes[left].GiftID < scopes[right].GiftID
	})
	return scopes
}

func compareBlindBoxViewers(left, right viewerContribution) int {
	if left.BlindBoxProfit != right.BlindBoxProfit {
		if left.BlindBoxProfit > right.BlindBoxProfit {
			return -1
		}
		return 1
	}
	if left.BlindBoxValue != right.BlindBoxValue {
		if left.BlindBoxValue > right.BlindBoxValue {
			return -1
		}
		return 1
	}
	if left.BlindBoxCount != right.BlindBoxCount {
		if left.BlindBoxCount > right.BlindBoxCount {
			return -1
		}
		return 1
	}
	if left.LastGiftAt != right.LastGiftAt {
		if left.LastGiftAt > right.LastGiftAt {
			return -1
		}
		return 1
	}
	return 0
}

func normalizeBlindBoxAmount(value float64) float64 {
	if math.IsNaN(value) || math.IsInf(value, 0) || value < 0 {
		return 0
	}
	return value
}

func normalizeBlindBoxCount(value int) int {
	return max(0, value)
}

func normalizeBlindBoxTimestamp(value int64) int64 {
	return max(int64(0), value)
}
